src/main/java/com/radixdlt/atomos/ParticleDefinition.java

```diff
@@ -18,35 +18,40 @@
 package com.radixdlt.atomos;
 
 import com.radixdlt.constraintmachine.Particle;
+import com.radixdlt.constraintmachine.Spin;
 import com.radixdlt.identifiers.RRI;
 import com.radixdlt.identifiers.RadixAddress;
 
+import java.util.Collections;
+import java.util.Set;
 import java.util.function.Function;
-import java.util.stream.Stream;
 
 /**
  * Defines how to retrieve important properties from a given particle type.
  * @param <T> the particle class
  */
-class ParticleDefinition<T extends Particle> {
-	private final Function<T, Stream<RadixAddress>> addressMapper;
-	private final Function<T, Result> staticValidation;
-	private final Function<T, RRI> rriMapper;
+public class ParticleDefinition<T extends Particle> {
+	private final Function<T, Set<RadixAddress>> addressMapper; // must be set (since we need to route the particle)
+	private final Function<T, Result> staticValidation; // may be null
+	private final Function<T, RRI> rriMapper; // may be null
+	private final Function<T, Spin> virtualizeSpin; // may be null
 	private final boolean allowsTransitionsFromOutsideScrypts;
 
-	ParticleDefinition(
-		Function<T, Stream<RadixAddress>> addressMapper,
+	private ParticleDefinition(
+		Function<T, Set<RadixAddress>> addressMapper,
 		Function<T, Result> staticValidation,
 		Function<T, RRI> rriMapper,
+		Function<T, Spin> virtualizeSpin,
 		boolean allowsTransitionsFromOutsideScrypts
 	) {
 		this.staticValidation = staticValidation;
 		this.addressMapper = addressMapper;
 		this.rriMapper = rriMapper;
+		this.virtualizeSpin = virtualizeSpin;
 		this.allowsTransitionsFromOutsideScrypts = allowsTransitionsFromOutsideScrypts;
 	}
 
-	Function<T, Stream<RadixAddress>> getAddressMapper() {
+	Function<T, Set<RadixAddress>> getAddressMapper() {
```

Comment:
Not really sure why this changed from a Stream to a Set?

Reply:
Easier to work with downstream since I need to check the size before accessing for validation